### PR TITLE
[pylint] Implement misplaced-bare-raise (E0704)

### PR DESCRIPTION
**Pull request by clemux**
Related issue: #970 
 
### What it does
This rule triggers an error when a bare raise statement is not in an except or finally block.
### Why is this bad?
If raise statement is not in an except or finally block, there is no active exception to
re-raise, so it will fail with a `RuntimeError` exception.
### Example
```python
def validate_positive(x):
   if x <= 0:
       raise
```
Use instead:
```python
def validate_positive(x):
   if x <= 0:
       raise ValueError(f"{x} is not positive")
```

## Test Plan

Added unit test and snapshot.
Manually compared ruff and pylint outputs on pylint's tests.

## References

 - [pylint documentation](https://pylint.pycqa.org/en/stable/user_guide/messages/error/misplaced-bare-raise.html)
 - [pylint implementation](https://github.com/pylint-dev/pylint/blob/main/pylint/checkers/exceptions.py#L339)

### CODE REVIEW

**Comment by charliermarsh** [comment]
Thanks! Do you mind taking a look at some of the violations flagged in the ecosystem check, and seeing how they compare to pylint's behavior?

**Comment by clemux** [comment]
They are the same with pylint. I had to run pylint on each file though, I'm not sure what I'm doing wrong. For example, with pip:

```
(venv) clemux@Nardole:~/dev/pip/src$ pylint --disable=all --enable=misplaced-bare-raise .
(venv) clemux@Nardole:~/dev/pip/src$ pylint --disable=all --enable=misplaced-bare-raise pip/_internal/utils/misc.py
************* Module pip._internal.utils.misc
pip/_internal/utils/misc.py:157:4: E0704: The raise statement is not inside an except clause (misplaced-bare-raise)

------------------------------------------------------------------
Your code has been rated at 9.85/10 (previous run: 9.85/10, +0.00)
```

File: crates/ruff_linter/src/rules/pylint/rules/misplaced_bare_raise.rs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    'b: 'a,
+{
+    fn visit_stmt(&mut self, stmt: &'b Stmt) {
+        match stmt {
+            Stmt::Raise(_) => self.raises.push(stmt),
+            _ => statement_visitor::walk_stmt(self, stmt),
+        }
+    }
+}
+
```
Comment (clemux) [inline crates/ruff_linter/src/rules/pylint/rules/misplaced_bare_raise.rs:58]: I wanted to use [`ruff_python_ast::helpers::RaiseStatementVisitor`](https://github.com/astral-sh/ruff/blob/main/crates/ruff_python_ast/src/helpers.rs#L895), but it doesn't store the raise statement itself.

It's only used in two rules, so maybe I can refactor that?

Comment (charliermarsh) [reply crates/ruff_linter/src/rules/pylint/rules/misplaced_bare_raise.rs:58]: Ah yeah -- can we change `RaiseStatementVisitor` to store `Vec<&'a ast::StmtRaise>`?

Comment (clemux) [reply crates/ruff_linter/src/rules/pylint/rules/misplaced_bare_raise.rs:58]: Shall I do that in a separate PR?

File: crates/ruff_linter/src/rules/pylint/rules/misplaced_bare_raise.rs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+                    let mut raise_finder = RaiseFinder::default();
+                    for handler in &try_stmt.handlers {
+                        raise_finder.visit_except_handler(handler);
+                    }
+                    if raise_finder.raises.contains(&raise_stmt) {
+                        return;
+                    }
+
+                    // if the raise is in the finally block, it might be ok if that finally is in
+                    // an except block
```
Comment (charliermarsh) [inline crates/ruff_linter/src/rules/pylint/rules/misplaced_bare_raise.rs:76]: Can you include a code example here in the comments to demonstrate the case that this is handling?

File: crates/ruff_linter/src/rules/pylint/rules/misplaced_bare_raise.rs (hunk 1 above)
Comment (charliermarsh) [reply crates/ruff_linter/src/rules/pylint/rules/misplaced_bare_raise.rs:58]: I would be cool with including it here, but either is fine.

File: crates/ruff_linter/src/rules/pylint/rules/misplaced_bare_raise.rs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+                        if fd.name.as_str() == "__exit__" {
+                            return;
+                        }
+                    }
+                    break;
+                }
+                _ => {}
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (clemux) [reply crates/ruff_linter/src/rules/pylint/rules/misplaced_bare_raise.rs:57]: Well... see the new diff :)
Thanks!

File: crates/ruff_linter/src/rules/pylint/rules/misplaced_bare_raise.rs (hunk 1 above)
Comment (clemux) [reply crates/ruff_linter/src/rules/pylint/rules/misplaced_bare_raise.rs]: Not needed anymore, so I guess I won't touch `RaiseStatementVisitor`.

File: crates/ruff_linter/src/rules/pylint/rules/misplaced_bare_raise.rs (hunk 2 above)
Comment (clemux) [reply crates/ruff_linter/src/rules/pylint/rules/misplaced_bare_raise.rs]: Not needed anymore!

**Pull request merged**